examples/table.rs

```diff
@@ -141,16 +183,16 @@ fn ui(f: &mut Frame, app: &mut App) {
             .cloned()
             .map(|content| Cell::from(Text::from(format!("\n{}\n", content))))
             .collect::<Row>()
-            .style(Style::new().bg(color))
+            .style(Style::new().fg(row_fg).bg(color))
             .height(4)
     });
     let bar = " █ ";
     let t = Table::new(
         rows,
         [
-            Constraint::Length(15),
+            Constraint::Length(25),
+            Constraint::Min(40),
             Constraint::Min(30),
-            Constraint::Min(25),
         ],
```

Comment:
It's probably worth calculating the maximum item lengths at least for the name and address here, and then letting the email take up the rest.

Reply:
I wrote the calculation function and added + 1 to all of the lengths before sending them `Constraint::Length `because they were looking too close.

```
    let (longest_name_len, longest_address_len, longest_email_len) =
        constraint_len_calculator(&app.items);
    let t = Table::new(
        rows,
        [
            // + 1 is for padding.
            Constraint::Length(longest_name_len + 1),
            Constraint::Min(longest_address_len + 1),
            Constraint::Min(longest_email_len + 1),
        ],
```

```
fn constraint_len_calculator(items: &[Vec<String>]) -> (u16, u16, u16) {
    let len_vec = items
        .iter()
        .map(|row| {
            (
                row[0].chars().count() as u16,
                row[1]
                    .split('\n')
                    .map(|address_line| address_line.chars().count())
                    .max()
                    .unwrap() as u16,
                row[2].chars().count() as u16,
            )
        })
        .collect::<Vec<(u16, u16, u16)>>();

    let longest_name_len = len_vec.iter().map(|row| row.0).max().unwrap();
    let longest_address_len = len_vec.iter().map(|row| row.1).max().unwrap();
    let longest_email_len = len_vec.iter().map(|row| row.2).max().unwrap();

    (longest_name_len, longest_address_len, longest_email_len)
}
```

Reply:
I can add a test function for `fn constraint_len_calculator()` and I'm open to name suggestions :D

Reply:
What about adding a small struct that has name, address, email and which gets generated in the app constructor, and then also calculating the column width values there and storing them in an app field rather than in each render?

This swaps indexing the columns implicitly for
```rust
let name_len =
  people.map(Person::name)
    .map(UnicodeWidth::width)
    .max();
let address_len
  people.map(Person::address)
    .flatmap(String::lines)
    .map(UnicodeWidth::width)
    .max();
...

```

Obviously this is just gold plating things at this point. But it does give a nice way of showing how to map real world data into table columns.